xc7/fasm2bels/cmt_models.py

```diff
@@ -0,0 +1,164 @@
+
+    for clkout in clkouts:
+        if site.has_feature('CLK{}_CLKOUT1_OUTPUT_ENABLE'.format(clkout)):
+
+            # Add output source
+            site.add_source(pll, 'CLK' + clkout, 'CLK' + clkout)
+
+            # Calculate the divider and duty cycle
+            high_time = decode_multi_bit_feature(
+                features, 'CLK{}_CLKOUT1_HIGH_TIME'.format(clkout)
+            )
+            low_time = decode_multi_bit_feature(
+                features, 'CLK{}_CLKOUT1_LOW_TIME'.format(clkout)
+            )
+
+            if decode_multi_bit_feature(features,
+                                        'CLK{}_CLKOUT2_EDGE'.format(clkout)):
+                high_time += 0.5
+                low_time = max(0, low_time - 0.5)
+
+            divider = int(high_time + low_time)
+            duty = high_time / (low_time + high_time)
+
+            if site.has_feature('CLK{}_CLKOUT2_NO_COUNT'.format(clkout)):
+                divider = 1
+                duty = 0.5
+
+            if clkout == 'FBOUT':
+                pll.parameters['CLKFBOUT_MULT'] = divider
+            else:
+                pll.parameters['CLK{}_DIVIDE'.format(clkout)] = divider
+                pll.parameters['CLK{}_DUTY_CYCLE'.format(clkout)
+                               ] = "{0:.3f}".format(duty)
+
+            # Phase shift
+            delay = decode_multi_bit_feature(
+                features, 'CLK{}_CLKOUT2_DELAY_TIME'.format(clkout)
+            )
+            phase = decode_multi_bit_feature(
+                features, 'CLK{}_CLKOUT1_PHASE_MUX'.format(clkout)
+            )
+
+            phase = float(delay) + phase / 8.0  # Delay in VCO cycles
+            phase = 360.0 * phase / divider  # Phase of CLK in degrees
+
+            if clkout == 'FBOUT':
+                pll.parameters['CLKFBOUT_PHASE'] = "{0:.3f}".format(phase)
+            else:
+                pll.parameters['CLK{}_PHASE'.format(clkout)
+                               ] = "{0:.3f}".format(phase)
+
+    # Input clock divider
+    high_time = decode_multi_bit_feature(features, 'DIVCLK_DIVCLK_HIGH_TIME')
+    low_time = decode_multi_bit_feature(features, 'DIVCLK_DIVCLK_LOW_TIME')
+
+    divider = high_time + low_time
+
+    if site.has_feature('DIVCLK_DIVCLK_NO_COUNT'):
+        divider = 1
+
+    pll.parameters['DIVCLK_DIVIDE'] = divider
+
+    # Startup wait
+    pll.parameters['STARTUP_WAIT'] = '"TRUE"' if site.has_feature(
+        'STARTUP_WAIT'
+    ) else '"FALSE"'
+
+    # Compensation  TODO: Other modes
+    if site.has_feature('COMPENSATION.INTERNAL'):
+        pll.parameters['COMPENSATION'] = '"INTERNAL"'
+
+    # Built-in inverters
+    pll.parameters['IS_CLKINSEL_INVERTED'] =\
+        "1'b1" if site.has_feature('INV_CLKINSEL') else "1'b0"
+    pll.parameters['IS_PWRDWN_INVERTED'] =\
+        "1'b1" if site.has_feature('ZINV_PWRDWN') else "1'b0"
+    pll.parameters['IS_RST_INVERTED'] =\
+        "1'b1" if site.has_feature('ZINV_RST') else "1'b0"
```

Comment:
Missing parameter for BANDWIDTH, which is determined (somehow) via the FILT regs.  Might need to a make a table, which would also be useful when going from verilog -> eblif -> VPR.

Reply:
Done, but it seems that "OPTIMIZED" and "HIGH" are the same.